pkg/settings/setting.go

```diff
@@ -117,45 +117,106 @@ type NonMaskedSetting interface {
-//     TenantWritable setting.
+//     1. if an end-user should be able to modify the setting in
+//     CockroachCloud Serverless, AND different end-users should be
+//     able to use different values, the setting should have class
+//     ApplicationLevel.
 //
-//   - Control settings relevant to tenant-specific internal implementation
-//     should be TenantReadOnly.
+//     2. if a setting should only be controlled by SREs in
+//     CockroachCloud Serverless, OR a single value must apply to
+//     multiple tenants (virtual clusters) simultaneously, the setting
+//     must not use class ApplicationLevel.
 //
-//   - When in doubt, the first choice to consider should be TenantReadOnly.
+//     3. if and only if a setting relevant to the KV/storage layer
+//     and whose value is shared across all virtual cluster is ever
+//     accessed by code in the application layer (SQL execution or
+//     HTTP handlers), use SystemVisible.
 //
-//   - SystemOnly should be used with caution: even internal tenant code is
-//     disallowed from using these settings at all.
+//     4. in other cases, use SystemOnly.
 type Class int8
 
 const (
-	// SystemOnly settings are associated with single-tenant clusters and host
-	// clusters. Settings with this class do not exist on non-system tenants and
-	// can only be used by the system tenant.
+	// SystemOnly settings are specific to the KV/storage layer and
+	// cannot be accessed from application layer code (in particular not
+	// from SQL layer nor HTTP handlers).
+	//
+	// As a rule of thumb, use this class if:
+	//
+	//   - the setting may be accessed from the shared KV/storage layer;
+	//
+	//   - AND, the setting should only be controllable by SREs (not
+	//     end-users) in CockroachCloud Serverless, AND a single value
+	//     must apply to all virtual clusters simultaneously;
+	//
+	//     (If this part of the condition does not hold, consider
+	//     ApplicationLevel instead.)
+	//
+	//   - AND, its value is never needed in the application layer (i.e.
+	//     it is not read from SQL code, HTTP handlers or other code
+	//     that would run in SQL pods in CC Serverless).
+	//
+	//     (If this part of the condition does not hold, consider
+	//     SystemVisible instead.)
 	SystemOnly Class = iota
 
-	// TenantReadOnly settings are visible to non-system tenants but cannot be
-	// modified by the tenant. Values for these settings are set from the system
-	// tenant and propagated from the host cluster.
-	TenantReadOnly
+	// SystemVisible settings are specific to the KV/storage layer and
+	// are also visible to virtual clusters.
+	//
+	// As a rule of thumb, use this class if:
+	//
+	//   - the setting may be accessed from the shared KV/storage layer;
+	//
+	//   - AND the setting should only be controllable by SREs (not
+	//     end-users) in CockroachCloud Serverless, AND a single value
+	//     must apply to all virtual clusters simultaneously;
```

Comment:
We have some settings like those in the tenantcostclient that sort of violate this as written.  In the current documentation, it looks like we should have made those ApplicationLevel and CC should have set an override.

Reply:
Very astute observation. I am very keen to have that convo with the CC team. But I feel that this convo will only be productive after we settle on the guidance.

Reply:
I agree. I don't think we should rely on external process to set those overrides though - we could set them ourselves by default (in a permanent upgrade).